### DISCUSSION TITLE
Force Discharge Period Investigation

### DISCUSSION THREAD

**Discussion by canton7** [nathanmarlor/foxess_modbus#440] [General]
It looks like force discharge periods have arrived!

This discussion is a place for us to figure out how they're exposed over modbus.

Annoyingly I'm on a super busy project at work at the moment, but I should have the odd day here and there to look at things.

**Comment by nathanmarlor** [comment]
Done some light digging this morning after updating to v1.70. Looks like Fox have added a whole new register set hidden behind a feature toggle. The existing charge periods still exist:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/3436911/64382373-eddc-41b1-bbba-01bca187e32c)

The new feature toggle:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/3436911/81eebee8-6438-4424-98fd-0eb688cd0da8)

The time periods are a similar structure to the existing, with a couple of bonus fields:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/3436911/4e712a01-2ad8-4012-a7a0-2b6eb3a27329)

For additional reference, Fox have published their API info -> https://www.foxesscloud.com/i18n/en/OpenApiDocument.html

The new fields are described as:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/3436911/3f6be93c-3676-41f9-bf3c-5cd602c5987c)

**Reply by TonyM1958** [reply]
If you try this, you may find you get a parameter error if you send 'fdSoc' or 'fdPwr' when the 'workMode' is not 'ForceDischarge' even though, when you get the schedule, the API now returns these fields (set to 0) in the data for other modes.

minSocOnGrid also may give parameter error - although it says integer here, the API only seems to accept a string in the JSON. It also seems to work with 'soc' instead of 'minSocOnGrid'.

**Reply by TonyM1958** [reply]
Looks like the problem with soc only accepting strings was fixed when the cloud API was updated to work with Master 1.70. minOnGridSoc now accepts integer as well as string values in the JSON payload.

**Comment by nathanmarlor** [comment]
Looks like the register addresses are sat at 46000+

```
([46010, 46011, 46012, 46013, 46014, 46015, 46016, 46017, 46018, 46019, 46020, 46021, 46022, 46023, 46024, 46025, 46026, 46027, 46028, 46029, 46030, 46031, 46032, 46033, 46034, 46035, 46036, 46037, 46038, 46039, 46040, 46041, 46042, 46043, 46044, 46045, 46046, 46047, 46048, 46049, 46050, 46051, 46052, 46053, 46054, 46055, 46056, 46057, 46058, 46059, 46060, 46061, 46062, 46063, 46064, 46065, 46066, 46067, 46068, 46069, 46070, 46071, 46072, 46073, 46074, 46075, 46076, 46077, 46078, 46079, 46080, 46081, 46082, 46083, 46084, 46085, 46086, 46087, 46088, 46089], 

enable / start / end / work mode / min_soc / fdsoc / fdpwr / ? / ? / ?
[1, 0, 1311, 6, 10, 10, 0, 0, 0, 0, 
1, 1311, 5662, 1, 10, 10, 0, 0, 0, 0, 
1, 5662, 5918, 7, 10, 20, 3600, 0, 0, 0, 
1, 5918, 5947, 6, 10, 10, 0, 0, 0, 0, 
0, 0, 0, 0, 10, 10, 0, 0, 0, 0, 
0, 0, 0, 0, 10, 10, 0, 0, 0, 0, 
0, 0, 0, 0, 10, 10, 0, 0, 0, 0, 
0, 0, 0, 0, 10, 10, 0, 0, 0, 0]))

```

... 8 periods to be set.

**Reply by TonyM1958** [reply]
So, it looks like 6 = ForceCharge and 7 = ForceDischarge work mode codes?

**Reply by TonyM1958** [reply]
BTW, kind of obvious - but for the record... ForceCharge work mode = Charge from Grid, not what we know as Force Charge. Just to further confuse the end users!

You can send fdPwr and fdSoc when setting ForceCharge but they don't seem to do anything useful. Would have been nice and maybe logical if fdPwr could set the Max input power and fdSoc could set Max Soc...

**Reply by TonyM1958** [reply]
@nathanmarlor - I think part of the trick to this will be the TimeModeFlag referenced above - looks like this is the overall Enable / Disable toggle that sets the inverter in strategic mode / normal mode. Looking at the 8 x 10 registers above, these are the period settings.

Did you find a modbus address for the TimeModeFlag?

Even if you can't load the periods, it would be good to be able to see what mode the inverter is running in based on the TimeModeFlag as this might tell you if you are able to update charge settings or work mode at that time.

**Reply by nathanmarlor** [reply]
@TonyM1958 iirc it was sat at 46000. I can double check. As with the others it's inaccessible through the rs485 interface.

**Reply by TonyM1958** [reply]
OK - another interesting one - when you are running a schedule, register 41000 doesn't update to reflect the work mode schedule. For example, I just ran a 15 minute scheduled ForceCharge and the work mode remained SelfUse throughout. Previously, it didn't change when you charged through the battery settings so it could be argued this isn't unexpected - but I did hope there would be some externally accessible indicator of what the inverter was doing?

**Comment by nathanmarlor** [comment]
Unfortunately it looks like these are inaccessible through the RS485 interface (same as battery voltages):

![image](https://github.com/nathanmarlor/foxess_modbus/assets/3436911/ae10ed72-634c-4738-aaf4-98269c654b57)

I'm timing out, but would be good if someone could try this address space with a different adapter.

**Reply by nathanmarlor** [reply]
FYI - when Fox Cloud reads these registers it does it as a group of 80:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/3436911/d38d6633-5472-4bc1-a26a-5c08e87ad93b)

**Reply by FozzieUK** [reply]
I also tried the RS485 registers (admittedly with another Waveshare) and got the same reply timeout error - I tried both input and holding registers - reply timeout.
![image](https://github.com/nathanmarlor/foxess_modbus/assets/113460294/33a6bf56-0d16-43b9-939e-ad25f143e901)

**Comment by spmaid** [comment]
> Unfortunately it looks like these are inaccessible through the RS485 interface (same as battery voltages):
> 
> ![image](https://user-images.githubusercontent.com/3436911/276241628-ae10ed72-634c-4738-aaf4-98269c654b57.png)
> 
> I'm timing out, but would be good if someone could try this address space with a different adapter.

I have LAN access to my inverter, is there anything I can help with testing?

**Reply by spmaid** [reply]
I now no longer have LAN access after upgrading myself to 1.70 to get the forced discharge, had hoped that would come back, but I believe accessing over RS485 is better anyway as offers more options?

**Reply by TonyM1958** [reply]
Fox disabled LAN port in 1.70 master firmware. You now can only get modbus data over RS485. New inverters don't even have the LAN port.

**Reply by spmaid** [reply]
Thank you for confirming!

**Comment by Ndrinta** [comment]
Hi, just updated to 1.70 and gave it a go. Same result in here too.
Anyone has any news about this?

**Reply by TonyM1958** [reply]
no response from the inverter over modbus either. Only current working solution to programming schedules is via the Fox Cloud API so far as I know.

**Reply by iancg** [reply]
Do you know if there is an integration which supports setting schedules via the Fox Cloud API - ideally one I can use in conjunction with this one (i.e. to use this one 99% of the time, but just program short periods of force discharge via the Cloud API)?  Could this integration have the limited cloud functionality included to get & set/enable/disable the schedules?

**Reply by TonyM1958** [reply]
This library does exactly this using the fox cloud API: https://github.com/TonyM1958/FoxESS-Cloud

I run it via Colab, Jupyter Lab or Saturn Cloud as described in the wiki.

One of the USPs of Nathan's integration is that it does not depend on Fox Cloud availability, so he will may comment on adding Cloud dependency into his integration...? 🤷‍♂️

**Comment by garethmarlow** [comment]
I can't code at this level unfortunately, but anything I can do to help test/develop this functionality? FoxESS did a quick turn-around to upgrade my firmware ahead of yesterday's saving sessions and it would be awesome to be able to set strategy periods programmatically without cloud access for future force-discharge events.

**Reply by canton7** [reply]
At the moment, it seems to be an inverter limitation, rather than anything that's missing on our side.

**Comment by canton7** [comment]
See #513 -- it looks like we can control force discharge using the remote control registers. (And always have been able to...)

**Reply by garethmarlow** [reply]
Good news. Happy to help test.
